hazelcast/src/main/java/com/hazelcast/jet/impl/JetClientInstanceImpl.java

```diff
@@ -149,96 +143,89 @@ public Job newJobProxy(long jobId, boolean isLightJob, @Nonnull Object jobDefini
                                     SubmitJobParameters submitJobParameters) {
-        ClientMessage jobMetaDataRequest = JetUploadJobMetaDataCodec.encodeRequest(sessionId,
-                fileNameWithoutExtension, sha256Hex,
+        ClientMessage jobMetaDataRequest = JetUploadJobMetaDataCodec.encodeRequest(
+                jobUploadCall.getSessionId(),
+                jobUploadCall.getFileNameWithoutExtension(),
+                jobUploadCall.getSha256HexOfJar(),
                 submitJobParameters.getSnapshotName(),
                 submitJobParameters.getJobName(),
                 submitJobParameters.getMainClass(),
                 submitJobParameters.getJobParameters());
 
-        return invokeRequestOnMasterAndDecodeResponse(jobMetaDataRequest, JetUploadJobMetaDataCodec::decodeResponse);
+        return invokeRequestAndDecodeResponseNoRetryOnRandom(jobUploadCall.getMemberUuid(), jobMetaDataRequest,
+                JetUploadJobMetaDataCodec::decodeResponse);
     }
 
-    private void sendJobMultipart(Path jarPath, UUID sessionId, long jarSize)
+    private void sendJobMultipart(JobUploadCall jobUploadCall, Path jarPath)
             throws IOException, NoSuchAlgorithmException {
-        int partSize = calculatePartBufferSize();
 
-        int totalParts = calculateTotalParts(jarSize, partSize);
 
         File file = jarPath.toFile();
 
-        byte[] data = new byte[partSize];
+        byte[] partBuffer = jobUploadCall.allocatePartBuffer();
 
         try (FileInputStream fileInputStream = new FileInputStream(file)) {
 
-            for (int currentPartNumber = 1; currentPartNumber <= totalParts; currentPartNumber++) {
+            // Start from part #1
+            for (int currentPartNumber = 1; currentPartNumber <= jobUploadCall.getTotalParts(); currentPartNumber++) {
+
+                int bytesRead = fileInputStream.read(partBuffer);
 
-                // Read data
-                int bytesRead = fileInputStream.read(data);
+                byte[] dataToSend = jobUploadCall.getDataToSend(partBuffer, bytesRead);
 
-                String sha256Hex = Sha256Util.calculateSha256Hex(data, bytesRead);
+                String sha256Hex = Sha256Util.calculateSha256Hex(dataToSend);
                 //Send the part
-                ClientMessage jobDataRequest = JetUploadJobMultipartCodec.encodeRequest(sessionId, currentPartNumber,
-                        totalParts, data, bytesRead, sha256Hex);
-                boolean result = invokeRequestOnMasterAndDecodeResponse(jobDataRequest,
+                ClientMessage jobMultipartRequest = JetUploadJobMultipartCodec.encodeRequest(
+                        jobUploadCall.getSessionId(),
+                        currentPartNumber,
+                        jobUploadCall.getTotalParts(), dataToSend, bytesRead, sha256Hex);
+
+                boolean result = invokeRequestAndDecodeResponseNoRetryOnRandom(jobUploadCall.getMemberUuid(), jobMultipartRequest,
```

Comment:
Yes. Because the upload is a multipart operation. Even if we try on a random member, the previous parts may end up on a different member and then the operation will fail anyway. So all parts should go to the same member in all or none manner.